src/mbgl/style/expression/type_check.cpp

```diff
@@ -0,0 +1,121 @@
+#include <mbgl/style/expression/type_check.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+using namespace mbgl::style::expression::type;
+
+TypecheckResult typecheck(const Type& expected, const std::unique_ptr<Expression>& e) {
```

Comment:
Sketch of possible definition syntax in JS:

```
    RegisterCompoundExpression('==', [
        [BooleanType, NullType, NullType],
        () => `true`,
        [BooleanType, NumberType, NumberType],
        (lhs, rhs) => `${lhs} === ${rhs}`,
        [BooleanType, StringType, StringType],
        (lhs, rhs) => `${lhs} === ${rhs}`,
        [BooleanType, BooleanType, BooleanType],
        (lhs, rhs) => `${lhs} === ${rhs}`,
        [BooleanType, ColorType, ColorType],
        (lhs, rhs) => `${lhs}.equal(${rhs})`
    ]),
```

In C++, you could just provide the evaluation methods, and the signatures could be deduced from those.